File: rest/src/main/java/com/tugab/adspartners/domain/entities/Ad.java
package com.tugab.adspartners.domain.entities;

import lombok.Data;
import org.hibernate.validator.constraints.Length;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;
import java.util.List;

@Data
@Entity
@Table(name = "ad")
public class Ad {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "description")
    private String description;

    @Column(name = "reward", nullable = false)
    private Double reward;

    @Column(name = "creation_date", nullable = false)
    private Date creationDate;

    @Column(name = "valid_to", nullable = false)
    private Date validTo;

    @Column(name = "min_videos")
    private Long minVideos;

    @Column(name = "min_subscribers")
    private Long minSubscribers;

    @Column(name = "min_views")
    private Long minViews;

    @Column(name = "is_blocked", nullable = false)
    private Boolean isBlocked;

    @ManyToOne(targetEntity = Company.class)
    @JoinColumn(name = "creator_id", referencedColumnName = "id")
    private Company company;

    @ManyToOne(targetEntity = CloudinaryResource.class, cascade = CascadeType.REMOVE)
    @JoinColumn(name = "picture_id", referencedColumnName = "id")
    private CloudinaryResource picture;

    @OneToMany(mappedBy = "ad", targetEntity = Characteristic.class, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Characteristic> characteristics;

    @OneToMany(mappedBy = "id.ad", targetEntity = AdRating.class, cascade = CascadeType.ALL)
    private List<AdRating> ratingList;

    @OneToMany(mappedBy = "id.ad", targetEntity = AdApplication.class, cascade = CascadeType.ALL)
    private List<AdApplication> applicationList;

    @Transient
    private Double averageRating = 0.0;
}

File: ui/src/validations/ad.js
export const title = value => {
    if (value.length < 3 || value.length > 256) {
        return 'Заглавието трябва да съдържа между 2 и 257 символа.';
    }

    return '';
};

export const shortDescription = value => {
    if (value.length > 1024) {
        return 'Описанието не трябва да надвищава 1024 символа.';
    }

    return '';
};

export const reward = value => {
    if (value === '') {
        return 'Възнаграждението е задължително.';
    }

    if (value < 0) {
        return 'Възнаграждението не трябва да бъде отрицателно.';
    }

    return '';
};

export const validTo = value => {
    if (value === '') {
        return 'Задължително е да предоставите крайна дата рекламата.';
    }

    if (new Date(value) - new Date() < 0) {
        return 'Датата трябва да бъде поне един ден валидна.';
    }

    return '';
};

export const minVideos = value => {
    if (value < 0) {
        return 'Броят видеа не може да бъде отрицателно число.';
    }

    return '';
};

export const minSubscribers = value => {
    if (value < 0) {
        return 'Броят абонати не може да бъде отрицателно число.';
    }

    return '';
};

export const minViews = value => {
    if (value < 0) {
        return 'Броят показвания не може да бъде отрицателно число.';
    }

    return '';
};

export const picture = value => {
    if (value === null) {
        return 'Задължително е да предоставите снимка на рекламата.';
    }

    return '';
};

export const charName = value => {
    if (value.length < 2) {
        return 'Името не може да бъде под 2 символа.';
    }

    return '';
};

export const charValue = value => {
    if (value.length === 0) {
        return 'Стойността не може да бъде празна.';
    }

    return '';
};
